scripts/nn/layers/attention.dml

```diff
@@ -0,0 +1,108 @@
+#-------------------------------------------------------------
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+#-------------------------------------------------------------
+
+source("nn/layers/softmax.dml") as softmax
+
+
+forward = function(matrix[double] query, matrix[double] key, matrix[double] value, integer K)
+    return (matrix[double] attention) {
+  /*
+   * Computes the forward pass for the attention layer.
+   *
+   * Inputs:
+   * - query: Input querys of shape (N,K*M).
+   * - key: Key(s) for value(s) of shape (N,K*M).
+   * - value: Value(s) for key(s) of shape (N,K*L).
+   * - K: Sequence length / number of timesteps.
+   * Outputs:
+   * - attention: Attention on value(s) for given query(s), of shape (N,K*L).
+   */
+  N = nrow(key)
+  M = ncol(query) / K
+  L = ncol(value) / K
+  norm = 1/M^0.5
+  key_norm = key * norm
+  attention = matrix(0, rows=N, cols=K*L)
+  for (n in 1:N)
+  {
+    query_n = matrix(query[n], rows=K, cols=M)
+    key_norm_n = matrix(key_norm[n],rows=K, cols=M)
+    value_n = matrix(value[n], rows=K, cols=L)
+    scores = query_n %*% t(key_norm_n)
```

Comment:
This is inefficient, you slice out a single row in each of the inputs and then matrix multiply.
instead you should simply multiply the entire input once to get scores.
with this change you can remove the entire for loop and process on matrices instead.

Reply:
We're having a little trouble with this... 
Since each line is a vectorized matrix, if we reshape and matrix multiply we get a lot of unused values.
The only values we need are then as blocks on the diagonal.

Reply:
we have a diagonal command to extract the diagonal from a matrix.

But if it is the case that you only need the diagonal,
then there should still be some smarter way of looking at it,
to still avoid this slicing of the matrix,
but it can also be that i was wrong in my analysis of the script.

Reply:
okay, looking closer, this is hard because of us treating the entire input as one matrix. If we had it slightly different we could have solved it with element wise multiplications and rowSums, but this is not feasible currently i will think about it a bit more since the current way (and the only way i can tell currently) still is inefficient.